datastore/postgres/affectedmanifest.go

```diff
@@ -18,336 +16,241 @@ import (
-//
-// The manifest index is then queried to resolve a list of manifest hashes containing the affected
-// artifacts.
-func (s *IndexerStore) AffectedManifests(ctx context.Context, v claircore.Vulnerability, vulnFunc claircore.CheckVulnernableFunc) ([]claircore.Digest, error) {
-	const (
-		selectPackages = `
-SELECT
-	id,
-	name,
-	version,
-	kind,
-	norm_kind,
-	norm_version,
-	module,
-	arch
-FROM
-	package
-WHERE
-	name = $1;
-`
-		selectAffected = `
-SELECT
-	manifest.hash
-FROM
-	manifest_index
-	JOIN manifest ON
-			manifest_index.manifest_id = manifest.id
-WHERE
-	package_id = $1
-	AND (
-			CASE
-			WHEN $2::INT8 IS NULL THEN dist_id IS NULL
-			ELSE dist_id = $2
-			END
-		)
-	AND (
-			CASE
-			WHEN $3::INT8 IS NULL THEN repo_id IS NULL
-			ELSE repo_id = $3
-			END
-		);
-`
-	)
-	ctx = zlog.ContextWithValues(ctx, "component", "datastore/postgres/affectedManifests")
-
-	// confirm the incoming vuln can be
-	// resolved into a prototype index record
-	pr, err := protoRecord(ctx, s.pool, v)
-	switch {
-	case err == nil:
-		// break out
-	case errors.Is(err, ErrNotIndexed):
-		// This is a common case: the system knows of a vulnerability but
-		// doesn't know of any manifests it could apply to.
-		return nil, nil
-	default:
-		return nil, err
-	}
-
-	// collect all packages which may be affected
-	// by the vulnerability in question.
-	pkgsToFilter := []claircore.Package{}
-
-	start := time.Now()
-	rows, err := s.pool.Query(ctx, selectPackages, v.Package.Name)
-	switch {
-	case errors.Is(err, nil):
-	case errors.Is(err, pgx.ErrNoRows):
-		return []claircore.Digest{}, nil
-	default:
-		return nil, fmt.Errorf("failed to query packages associated with vulnerability %q: %w", v.ID, err)
-	}
-	defer rows.Close()
-	affectedManifestsCounter.WithLabelValues("selectPackages").Add(1)
-	affectedManifestsDuration.WithLabelValues("selectPackages").Observe(time.Since(start).Seconds())
-
-	for rows.Next() {
-		var pkg claircore.Package
-		var id int64
-		var nKind *string
-		var nVer pgtype.Int4Array
-		err := rows.Scan(
-			&id,
-			&pkg.Name,
-			&pkg.Version,
-			&pkg.Kind,
-			&nKind,
-			&nVer,
-			&pkg.Module,
-			&pkg.Arch,
-		)
-		if err != nil {
-			return nil, fmt.Errorf("failed to scan package: %w", err)
+func queryBuilder(v *claircore.Vulnerability, vp claircore.CheckVulnerablePipeline) (string, error) {
+	psql := goqu.Dialect("postgres")
+	exps := []goqu.Expression{}
+	packageQuery := goqu.Ex{"package.name": v.Package.Name}
+	exps = append(exps, packageQuery)
+	seen := make(map[claircore.MatchConstraint]struct{})
+	for _, m := range vp.Query() {
+		if _, ok := seen[m]; ok {
+			continue
 		}
-		idStr := strconv.FormatInt(id, 10)
-		pkg.ID = idStr
-		if nKind != nil {
-			pkg.NormalizedVersion.Kind = *nKind
-			for i, n := range nVer.Elements {
-				pkg.NormalizedVersion.V[i] = n.Int
+		var ex goqu.Ex
+		switch m {
+		case claircore.PackageModule:
+			ex = goqu.Ex{"package.module": v.Package.Module}
```

Comment:
Needs a better explanation but essentially the RHEL matcher is doing something weird where it checks the package.module (and the package module is an empty string, all the other Matchers behave as you'd expect, ask to query on a column that actually have information about.